File: SecondSemester/OOP/Lab6/C++/Expression.cpp
//
// Created by bohdan on 27.04.20.
//

#include <stdexcept>
#include "Expression.h"
#include "cmath"

using namespace std;

Expression::Expression(double a, double b, double c) {
    set_A(a);
    set_B(b);
    set_C(c);
}

void Expression::Calculate() {
    if ((this->_a * this->_b + 2) <= 0) {
        throw std::invalid_argument("Log(a * b + 2) has only positive range of argument values");;
    }
    if ((41 - (this->_b / this->_c) + 1) == 0) {
        throw std::range_error("Division by zero is impossible!");;
    }
    this->_result = (log(this->_a * this->_b + 2) * 2) / (41 - (this->_b / this->_c) + 1);
}


double Expression::getResult() {
    return this->_result;
}

void Expression::set_A(double a) {
    this->_a = a;
}
void Expression::set_B(double b) {
    this->_b = b;
}
void Expression::set_C(double c) {
    if (c == 0) {
        throw std::invalid_argument("Division by zero has occurred!");
    }
    this->_c = c;
}

double Expression::get_A() {
    return this->_a;
}
double Expression::get_B() {
    return this->_b;
}
double Expression::get_C() {
    return this->_c;
}

File: SecondSemester/OOP/Lab6/C#/Expression.cs
using System;

namespace Lab6
{
    public class Expression
    {
        private double _result;

        private double _c;
        
        public double A { private get; set; }
        public double B { private get; set; }
        public double C {
            private get
            {
                return _c ;
            }
            set
            {
                if (value == 0d)
                {
                    throw new DivideByZeroException("Division by zero exception!");
                }
                _c = value;
            }
        }
        public Expression(double a, double b, double c)
        {
            A = a;
            B = b;
            C = c;
        }

        public void Calculate()
        {
            if ((A * B + 2) <= 0)
            {
                throw new ArgumentOutOfRangeException("A and B", "Log(a * b + 2) has only positive range of argument values");
            } 
            if (Math.Abs((41 - (B / C) + 1)) < Double.Epsilon)
            {
                throw new DivideByZeroException("Division by zero exception!");
            }
            _result = (Math.Log(A * B + 2) * 2) / (41 - (B / C) + 1);
        }

        public double GetResult()
        {
            return _result;

        }
    }
}